pkg/webhook/server_test.go

```diff
@@ -0,0 +1,108 @@
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	tassert "github.com/stretchr/testify/assert"
+
+	"github.com/openservicemesh/osm/pkg/certificate"
+	tresorFake "github.com/openservicemesh/osm/pkg/certificate/providers/tresor/fake"
+	"github.com/openservicemesh/osm/pkg/constants"
+)
+
+func TestCertRotation(t *testing.T) {
+	testCases := []struct {
+		name              string
+		rotations         int
+		rotationsExpected int
+		checkInterval     time.Duration
+		waitForRotation   bool
+	}{
+		{name: "should rotate certificate",
+			rotations:         1,
+			rotationsExpected: 1,
+			checkInterval:     10 * time.Millisecond,
+			waitForRotation:   true,
+		},
+		{name: "should rotate certificate if CA changes multiple times",
+			rotations:         2,
+			rotationsExpected: 2,
+			checkInterval:     10 * time.Millisecond,
+			waitForRotation:   true,
+		},
+		{name: "should not rotate certificate if no changes",
+			rotations:         0,
+			rotationsExpected: 0,
+			checkInterval:     10 * time.Millisecond,
+			waitForRotation:   true,
+		},
+		{name: "should not rotate certificate if checkInterval is longer than rotation",
+			rotations:         1,
+			rotationsExpected: 0,
+			checkInterval:     1 * time.Hour,
+			waitForRotation:   false,
+		},
+		{name: "should not rotate certificate if checkInterval is longer than rotation",
+			rotations:         2,
+			rotationsExpected: 0,
+			checkInterval:     1 * time.Hour,
+			waitForRotation:   false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert := tassert.New(t)
+
+			mrc := tresorFake.NewFakeMRC()
+			cm := tresorFake.NewFakeWithMRC(mrc, tc.checkInterval)
+
+			count := 0
+			wait := make(chan struct{})
```

Comment:
This could be simplified a tad, but not blocking on the review if you prefer to keep it.


Due to semantics of certmanager, you can "pre-issue" the certificate, by calling `firstCert, err := cm.IssueCertificate(...)`

That saves the use of needing to wait for the first onRotation to be called.

Now you can switch the chan stuff to a sync.WaitGroup. Makes things a bit more readable, but does depend on knowing the internals of certmanager. Just a thought!

Reply:
I think `onRotation` isn't named right here, it should be `onCertChange` (as the parameter is called in `NewServer`). 

Since this gets called for every instance, this if I pre-issue the cert it will this function is still called, but I tried out the waitgroup version and it is a bit simpler I think. Thanks!